index.rst

```diff
@@ -1,29 +1,559 @@
+  
+  Call ``vars(message)`` instead of ``message.get_vars()``.
+  Use of ``get_vars`` is quite rare, and easy to find.
+  Note that it is possible to call ``vars`` on DDS messages, but it returns one unwanted extra field: ``_member_attributes``.
+  
+* Explicitly encode bytes arrays for string-valued message fields.
+
+  OpenSplice DDS allows writing a bytes array as a string, encoding the data internally, but Kafka requires the data be explicitly encoded.
+  We do not expect to find many instances of this.
+  We do not recommend back-porting this change, because OpenSplice DDS does not support ``utf-8``, which is our preferred encoding.
+
+* Unit tests run much faster with Kafka.
+
+  As a result, poorly written tests may fail due to timing issues.
+  Such fixes can be back-ported.
+
+We converted three packages to work with Kafka ts_salobj: ``ts_atwhitelight``, ``ts_scriptqueue`` and ``ts_watcher``.
+The first is a simple CSC.
+The latter two have the most complex use of ts_salobj of all packages (except ``ts_salkafka``, which is not needed with Kafkfa ts_salobj).
+
+We only modified the code, not the build files, such as ``setup.cfg`` and ``setup.py``.
+
+Here is what we found:
+
+ts_atwhitelight
+~~~~~~~~~~~~~~~
+
+No changes required.
+
+ts_watcher
+~~~~~~~~~~
+
+Changes we have back-ported:
+
+* Some unit tests create ``SalIinfo`` instance, in order to write mock data from other CSCs.
+  Call ``await start()`` on those instances.
+
+Changes we prefer not to back-port:
+
+* One test called ``get_vars`` on a message; convert to ``vars``.
+
+* One test subclasses ts_salobj ``WriteTopic`` to write data with an offset value of ``private_sndStamp``.
+  This uses some internal details of ``WriteTopic`` that are different for the Kafka version of ts_salobj.
+
+ts_scriptqueue
+~~~~~~~~~~~~~~
+
+Changes we have back-ported:
+
+* Move ``super().start()`` to the top of ``ScriptQueue.start``.
+
+* Fix one test that had a race condition.
+
+Changes we provisionally plan to back-port:
+
+* Rename field ``ScriptID`` to ``salIndex`` if `RFC-849 <https://jira.lsstcorp.org/browse/RFC-849>`_ is accepted.
+
+Changes we prefer not to back-port:
+
+* Change one call from ``message.get_vars()`` to ``vars(message)``.
+
+* Fix one instance of writing a bytes array to a string field, by encoding the bytes array.
+
+Build Files
+-----------
+
+All packages will need build files updated, including:
+
+* Require Kafka-related packages instead of DDS-related packages.
+
+* Stop building IDL files in Jenkinsfile.
+  Technically this is optional, but this step is no longer necessary.
+  Kafka ts_salobj parses our ts_xml topic schema files directly.
+
+Converting ts_sal and C++, Java, and LabVIEW CSCs
+=================================================
+
+To port ts_sal from DDS to Kafka, the bulk of the effort would likely be in transitioning the topic message schema from OMG IDL to Avro.
+We would also expect to need to do some performance tuning to operate at scale.
```

Comment:
I agree and it may be best to just eliminate it. I frankly think it very unlikely. But Dave did bring it up.

Kafka tuning is more straightforward than DDS. There's less going on.

Reply:
I removed it